### PR TITLE
Use protobuf3 for internal serialization

### PR DESCRIPTION
**Pull request by chbatey**
Ignore all changes in the generated messages. I'll point out the important bits.

I have't converted the proto files yet, not sure that this is worth it and serves as documentation for which fields were mandatory for doing rolling upgrades with older akka versions.

Important changes:

* Uses sbt assembly shading via its own sub project. Package name is `akka.protobufv3.internal` to indicate to users not to use it
* All protobuf generated classes re-generated with `3.9.0`
* All serializers have been updated, only minor changes e.g. classes now extend GeneratedMessageV3
* Old protobuf left in in case users are using it

Will follow up with:
* Document that akka-protobuf is deprecated and should move to their own protobuf
* Document that akka-protobuf-v3 is internal

Testing has been done in https://github.com/lightbend/akka-upgrade-testing

This has isolated protobuf changes in proto2/proto3 projects as well as a test suite that uses minikube to upgrade akka across versions, validating that different akka versions can communicate with each other. I opted for a travis build as we can already have things running in minikube there and didn't want to have minikube tests on our jenkins.

### CODE REVIEW

**Comment by chbatey** [comment]
Looks like I’ve got some more sbt negotiation to do for the publish

File: akka-cluster-metrics/src/main/protobuf/ClusterMetricsMessages.proto [hunk 1]
```diff
@@ -2,6 +2,8 @@
  * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
  */
 
+syntax = "proto2";
```
Comment (chbatey) [inline akka-cluster-metrics/src/main/protobuf/ClusterMetricsMessages.proto]: protoc3 can still parse protoc2 files but the code generation will be new, not the old version

File: akka-cluster/src/main/protobuf/ClusterMessages.proto [hunk 2]
```diff
@@ -67,23 +69,26 @@ message ConfigCheck {
   optional string clusterConfig = 2;
 }
 
+// TODO check that making these fields optional is wire compatible
+// We must "always" set them otherwise a rolling update won't work
+// as older version deserialisation will fail
 message UncheckedConfig {
   extend ConfigCheck {
-    required UncheckedConfig checkConfig = 100;
+    optional UncheckedConfig checkConfig = 100;
```
Comment (chbatey) [inline akka-cluster/src/main/protobuf/ClusterMessages.proto]: I need to test that this is okay to do

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatedDataSerializer.scala [hunk 3]
```diff
@@ -8,27 +8,27 @@ import java.{ util, lang => jl }
 import java.util.ArrayList
 import java.util.Collections
 import java.util.Comparator
-import java.util.TreeSet
 
 import scala.annotation.tailrec
 import akka.util.ccompat.JavaConverters._
-import scala.collection.immutable
 
+import scala.collection.immutable
```
Comment (chbatey) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatedDataSerializer.scala]: changes to the replicated data serializer as it had generic methods for `GeneratedMessage` which is now `GeneratedMessageV3`

File: build.sbt [hunk 4]
```diff
@@ -62,6 +62,7 @@ lazy val aggregatedProjects: Seq[ProjectReference] = List[ProjectReference](
   persistenceTck,
   persistenceTyped,
   protobuf,
+  protobufV3,
```
Comment (chbatey) [inline build.sbt]: all modules should now depend on protobufV3

File: project/Protobuf.scala [hunk 5]
```diff
@@ -62,61 +84,77 @@ object Protobuf {
...
-      if (in.removed.nonEmpty || in.modified.nonEmpty) {
-        log.info("Preprocessing directory %s..." format sourceDir)
-        for (source <- in.removed; target <- map(source)) {
-          IO delete target
-        }
-        val updated = for (source <- in.modified; target <- map(source)) yield {
-          if (source.isFile) {
-            if (transformable(source)) transform(source, target)
-            else IO.copyFile(source, target)
+  private def transformDirectory(
```
Comment (chbatey) [inline project/Protobuf.scala]: intellij with format on save scalafmt is ver insistent on reformatting this, our build only checks source code is formatted not build files

File: akka-actor/src/main/scala/akka/serialization/Serialization.scala [hunk 6]
```diff
@@ -469,8 +469,14 @@ class Serialization(val system: ExtendedActorSystem) extends Extension {
   // i.e. com.google.protobuf dependency has been added in the application project.
   // The reason for this special case is for backwards compatibility so that we still can
   // include "com.google.protobuf.GeneratedMessage" = proto in configured serialization-bindings.
-  private def checkGoogleProtobuf(className: String): Boolean =
-    (!className.startsWith("com.google.protobuf") || system.dynamicAccess.getClassFor[Any](className).isSuccess)
+  private def checkGoogleProtobuf(className: String): Boolean = checkClass("com.google.protobuf", className)
+
+  // akka-protobuf is now not a dependency of remote so only load if user has explicitly added it
```
Comment (helena) [inline akka-actor/src/main/scala/akka/serialization/Serialization.scala]: 👍

File: akka-cluster-sharding-typed/src/main/protobuf/ShardingMessages.proto [hunk 7]
```diff
@@ -2,6 +2,8 @@
  * Copyright (C) 2014-2019 Lightbend Inc. <https://www.lightbend.com>
  */
 
+syntax = "proto2";
```
Comment (helena) [inline akka-cluster-sharding-typed/src/main/protobuf/ShardingMessages.proto]: Why proto2?

File: akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/TestMessage.scala [hunk 8]
```diff
@@ -7,7 +7,7 @@ package akka.remote.artery
 import akka.actor.ExtendedActorSystem
 import akka.serialization.SerializerWithStringManifest
 import akka.remote.artery.protobuf.{ TestMessages => proto }
-import akka.protobuf.ByteString
+import akka.protobufv3.internalByteString
```
Comment (helena) [inline akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/TestMessage.scala]: Is that supposed to be `import akka.protobufv3.internal.ByteString`?

**Review by helena — APPROVED** [review]
LGTM but caveat: I am not as familiar with akka proto internal requirements yet.

File: akka-cluster/src/main/protobuf/ClusterMessages.proto (hunk 2 above)
Comment (chbatey) [reply akka-cluster/src/main/protobuf/ClusterMessages.proto]: No need now, they are removed in another PR that'll we can merge before this

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala [hunk 9]
```diff
@@ -237,6 +239,7 @@ final class LmdbDurableStore(config: Config) extends Actor with ActorLogging {
 
   def active: Receive = {
     case Store(key, data, reply) =>
+      log.debug("Store key: {} reply: {} writeBehind {}", key, reply, writeBehindInterval.length)
```
Comment (chbatey) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala]: FIXME, remove, was debugging an issue in the repeat job

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala [hunk 10]
```diff
@@ -246,6 +249,7 @@ final class LmdbDurableStore(config: Config) extends Actor with ActorLogging {
             context.system.scheduler.scheduleOnce(writeBehindInterval, self, WriteBehind)(context.dispatcher)
           pending.put(key, data)
         }
+        log.debug("Write complete")
```
Comment (chbatey) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala]: FIXME, remove

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 11]
```diff
@@ -1403,6 +1403,8 @@ final class Replicator(settings: ReplicatorSettings) extends Actor with ActorLog
       val originalReplyTo = replyTo
       stash.foreach {
         case (msg, snd) =>
+          // FIXME remove
+          log.debug("Unstashing {} to with sender {}", msg, snd)
```
Comment (chbatey) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: FIXME ^

Comment (chbatey) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: For all the logs added in this file

Comment (helena) [reply akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: Confusing sentence.

File: akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingRememberEntitiesSpec.scala [hunk 12]
```diff
@@ -54,7 +54,7 @@ abstract class ClusterShardingRememberEntitiesSpecConfig(val mode: String) exten
   commonConfig(
     ConfigFactory
       .parseString(s"""
-    akka.loglevel = INFO
+    akka.loglevel = DEBUG
```
Comment (helena) [inline akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sharding/ClusterShardingRememberEntitiesSpec.scala]: Let's not keep that for CI long term

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala [hunk 13]
```diff
@@ -157,6 +157,7 @@ final class LmdbDurableStore(config: Config) extends Actor with ActorLogging {
           TimeUnit.NANOSECONDS.toMillis(System.nanoTime - t0))
       val l = Lmdb(env, db, keyBuffer, valueBuffer)
       _lmdb = OptionVal.Some(l)
+      log.debug("LMDB initialized")
```
Comment (helena) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala]: FIXME repeat job probably

**Comment by chbatey** [comment]
Don’t bother reviewing any logging I’ll remove it all before merge

File: akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/DurableDataSpec.scala [hunk 14]
```diff
@@ -23,7 +23,7 @@ final case class DurableDataSpecConfig(writeBehind: Boolean) extends MultiNodeCo
   val first = role("first")
   val second = role("second")
 
-  commonConfig(ConfigFactory.parseString(s"""
+  commonConfig(ConfigFactory.parseString(s"""akka.loglevel = DEBUG
```
Comment (helena) [inline akka-distributed-data/src/multi-jvm/scala/akka/cluster/ddata/DurableDataSpec.scala]: back to `INFO` eventually

File: akka-multi-node-testkit/src/main/mima-filters/2.5.x.backwards.excludes [hunk 15]
```diff
@@ -0,0 +1,3 @@
+# Protobuf 3
+ProblemFilters.exclude[Problem]("akka.remote.testconductor.TestConductorProtocol*")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.remote.testconductor.ProtobufDecoder.this")
```
Comment (helena) [inline akka-multi-node-testkit/src/main/mima-filters/2.5.x.backwards.excludes]: Do we need another constructor here?

**Review by helena — COMMENTED** [review]
Looking good, I guess this is still in flight with the newer debug/logging additions to troubleshoot the repeat CI job. Will look again when it's 💯

**Comment by raboof** [comment]
(rebased)

**Review by patriknw — APPROVED** [review]
LGTM, only some logging/fixme cleanups

**Comment by patriknw** [comment]
Refs #27366

**Comment by patriknw** [comment]
removed the debug logging

**Comment by raboof** [comment]
hmm `akka.remote.classic.RemotingSpec.Remoting must allow other system to connect even if it's not there at first` failed

**Comment by raboof** [comment]
PLS BUILD

**Comment by patriknw** [comment]
Merging, if that test fails again we can follow up

**Pull request merged**